### PR TITLE
[feature](function)support url domain functions

### PR DESCRIPTION
**Pull request by zhangstar333**
## Proposed changes

support top_level_domain/first_significant_subdomain/cut_to_first_significant_subdomain functions
doc: https://github.com/apache/doris-website/pull/1230

### CODE REVIEW

**Comment by LiBinfeng-01** [comment]
need to add fe fold const when adding new scalar function to fe, which can refer to: https://selectdb.feishu.cn/wiki/BGfGwgY2uiQrK7krGUVcYJFknVg?from=from_copylink

**Comment by zhangstar333** [comment]
run buildall

File: be/src/vec/functions/url/domain.h [hunk 1]
```diff
@@ -144,4 +145,138 @@ struct ExtractDomain {
     }
 };
 
+struct ExtractTopLevelDomain {
+    static size_t get_reserve_length_for_element() { return 5; }
+
+    static void execute(const char* data, size_t size, const char*& res_data, size_t& res_size) {
```
Comment (HappenLee) [inline be/src/vec/functions/url/domain.h:151]: why not return `StringRef`

**Comment by zhangstar333** [comment]
run buildall

File: be/src/vec/functions/url/find_symbols.h [hunk 2]
```diff
@@ -0,0 +1,484 @@
...
+// https://github.com/ClickHouse/ClickHouse/blob/master/base/base/find_symbols.h
+// and modified by Doris
+
+#pragma once
+
+#include <array>
+#include <cstdint>
+#include <string>
+
+#if defined(__SSE2__)
```
Comment (HappenLee) [inline be/src/vec/functions/url/find_symbols.h:27]: only define SSE4_2 enoungh. and here need add <sse2neno> header

File: be/src/vec/functions/url/domain.h [hunk 3]
```diff
@@ -144,4 +145,138 @@ struct ExtractDomain {
...
+            res_size = 0;
+        } else {
+            auto host_view = host.to_string_view();
+            if (host_view[host_view.size() - 1] == '.') {
+                host_view.remove_suffix(1);
+            }
+
+            const auto* host_end = host_view.data() + host_view.size();
+            const char* last_dot = find_last_symbols_or_null<'.'>(host_view.data(), host_end);
+            if (!last_dot) {
```
Comment (HappenLee) [inline be/src/vec/functions/url/domain.h:165]: if not find the `last_dot` what should the `res_data` and `res_size` ?

File: be/src/vec/functions/url/domain.h [hunk 4]
```diff
@@ -144,4 +145,138 @@ struct ExtractDomain {
...
+            }
+
+            res_data = last_dot + 1;
+            res_size = host_end - res_data;
+        }
+    }
+};
+
+struct FirstSignificantSubdomainDefaultLookup {
+    bool operator()(StringRef host) const { return tldLookup::isValid(host.data, host.size); }
```
Comment (HappenLee) [inline be/src/vec/functions/url/domain.h:184]: is_valid

File: be/src/vec/functions/url/domain.h [hunk 5]
```diff
@@ -144,4 +145,138 @@ struct ExtractDomain {
...
+};
+
+struct ExtractFirstSignificantSubdomain {
+    static size_t get_reserve_length_for_element() { return 10; }
+
+    static void execute(const Pos data, const size_t size, Pos& res_data, size_t& res_size,
+                        Pos* out_domain_end = nullptr) {
+        FirstSignificantSubdomainDefaultLookup loookup;
```
Comment (HappenLee) [inline be/src/vec/functions/url/domain.h:192]: why here we need a class, not direct call the static func `tldLooup::is_valid`

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/CutToFirstSignificantSubdomain.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+
+/**
+ * ScalarFunction 'CutToFirstSignificantSubdomain'. This class is generated by GenerateFunction.
+ */
+public class CutToFirstSignificantSubdomain extends ScalarFunction
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/CutToFirstSignificantSubdomain.java:36]: add constant folding in fe for these three functions

File: be/src/vec/functions/url/domain.h (hunk 1 above)
Comment (zhangstar333) [reply be/src/vec/functions/url/domain.h:151]: now the ExtractSubstringImpl function have abstract this function interface.
if changed maybe update other functions, it's a little more

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/CutToFirstSignificantSubdomain.java (hunk 6 above)
Comment (zhangstar333) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/CutToFirstSignificantSubdomain.java:36]: this functions need some more calculation algorithm, maybe not easy add in FE constant

**Comment by zhangstar333** [comment]
run buildall

**Review by HappenLee — APPROVED** [review]
LGTM

**Review by zclllyybb — APPROVED** [review]
LGTM

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/CutToFirstSignificantSubdomain.java [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+        implements UnaryExpression, ExplicitlyCastableSignature, PropagateNullable {
+
+    public static final List<FunctionSignature> SIGNATURES = ImmutableList.of(
+            FunctionSignature.ret(StringType.INSTANCE).args(StringType.INSTANCE)
```
[the remark this answers is not in the document]
Comment (zhangstar333) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/CutToFirstSignificantSubdomain.java:40]: same, could handle it

**Pull request merged**